### PR TITLE
members-only app

### PR DESCRIPTION
**Pull request by JayKowski**
ready for review

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    flash[:danger] = 'You must log in to perform that action'
+    redirect_to root_path
+  end
+
+  def create_token_and_save(user)
+    user.create_token_and_save
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:gen_token] = user.gen_token
+  end
```
Comment (wathigo) [inline app/controllers/application_controller.rb]: * We need to set the cookies every time the user logs in.
We can just update the `remeber_digest` attribute of the user and set a new token ccokie each time a user logs in.
The method should look similar ti this 
```ruby
def sign_in
  gen_token = User.new_token
  user.update_attribute(:remember_digest, User.encrypted(gen_token)
  cookies.permanent[:gen_token] = gen_token
  current_user = user
end
```

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      flash[:success] = 'You have successfully logged in'
+      redirect_to root_path
+    else
+      flash.now[:danger] = 'There is something wrong with your login information'
+      render 'new'
+    end
+  end
+
+  def destroy
+    session[:user_id] = nil
```
Comment (wathigo) [inline app/controllers/sessions_controller.rb]: Let's set cookies to nil each time the user logs out.
We can use a helper method for this that looks similar to.
```ruby
def sign_out
  @current_user = nil
  cookies.delete(:gen_token)
end
```

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+class ApplicationController < ActionController::Base
+  helper_method :current_user, :logged_in?
+
+  def current_user
+    @current_user ||= User.find(session[:user_id]) if session[:user_id]
```
Comment (wathigo) [inline app/controllers/application_controller.rb]: To remember a user we can use.
```
User.find_by(remember_digest: User.encrypted(cookies.permanent[:gen_token]))
```

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    user = User.find_by(email: params[:session][:email].downcase)
+    if user&.authenticate(params[:session][:password])
+      session[:user_id] = user.id
+      flash[:success] = 'You have successfully logged in'
+      redirect_to root_path
+    else
+      flash.now[:danger] = 'There is something wrong with your login information'
+      render 'new'
+    end
+  end
```
Comment (wathigo) [inline app/controllers/sessions_controller.rb]: We need to set a new `gen_token` in the cookies and update `remeber_digest` attribute each time a user logs in.
We can do this by invoking the `sign_in` helper method in the application controller(updated version of `create_token_and_save` in `controllers/application_controller` in our case. each time a user is successfully authenticated.

**Review by wathigo — CHANGES_REQUESTED** [review]
### Changes Required 
**Hi @JayKowski  and @Elijahscriptdev  :wave:**  

Great work so far :clap: 
However, the application is not able to remember logged in user after closing the browser.
We need to create a new token every time a user logs in and updated the remember digest attribute.
See the comments for suggestions :point_down: 

**Let's fix this and submit another request for review.**

**Happy Coding :computer: 
Simon Wathigo.**

**Review by imhta — APPROVED** [review]
# Project approved 🎉 

### Hello  @JayKowski, @Elijahscriptdev , 👋 You have done a great job in this project 👏 




----
### Happy coding 😄 
----
**Pon Muthu Selvam N**
__nponmuthuselvam@gmail.com__

**Comment by JayKowski** [comment]
Thankyou @imhta 😄😄😄👍👍👍💯💯💯

**Pull request merged**